src/osd/scrubber/scrub_backend.h

```diff
@@ -0,0 +1,419 @@
+// auth_selection_t listing in the logs
+template <>
+struct fmt::formatter<auth_selection_t> {
+  template <typename ParseContext>
+  constexpr auto parse(ParseContext& ctx)
+  {
+    return ctx.begin();
+  }
+
+  template <typename FormatContext>
+  auto format(auth_selection_t const& aus, FormatContext& ctx)
+  {
+    return fmt::format_to(ctx.out(),
+                          " {{AU-S: {}->{:x} OI({:x}:{}) {} dm:{}}} ",
+                          aus.auth->first,
+                          (uint64_t)(&aus.auth->second),
+                          (uint64_t)(&aus.auth_oi),
+                          aus.auth_oi,
+                          aus.shard_map.size(),
+                          aus.digest_match);
+  }
+};
+
+/**
+ * the back-end data that is per-chunk
+ *
+ * Created by the Scrubber after all replicas' maps have arrived.
+ */
+struct ScrubBeChunk {
+
+  explicit ScrubBeChunk(pg_shard_t i_am)
+  {
+    received_maps[i_am] = ScrubMap{};
+  }
+
+  /// the working set of scrub maps: the received maps, plus
+  /// Primary's own map.
+  std::map<pg_shard_t, ScrubMap> received_maps;
+
+  /// a collection of all objs mentioned in the maps
+  std::set<hobject_t> master_set;
+
+  utime_t started{ceph_clock_now()};
+
+  std::map<hobject_t, data_omap_digests_t> missing_digest;
+
+  /// Map from object with errors to good peers
+  std::map<hobject_t, std::list<pg_shard_t>> authoritative;
+
+
+  // these must be reset for each element:
+
+  std::set<pg_shard_t> cur_missing;
+  std::set<pg_shard_t> cur_inconsistent;
+  bool fix_digest{false};
+};
+
+
+/**
+ * ScrubBackend wraps the data and operations required for the back-end part of
+ * the scrubbing (i.e. for comparing the maps and fixing objects).
+ *
+ * Created anew upon each initiation of a scrub session.
+ *
+ * Uses 'optional' entries for winking-out the auxiliary collections.
+ */
+class ScrubBackend : public ScrubBackendIF {
+ public:
+  // Primary constructor
+  ScrubBackend(PgScrubber& scrubber,
+               PGBackend& backend,
+               PG& pg,
+               pg_shard_t i_am,
+               bool repair,
+               scrub_level_t shallow_or_deep,
+               const std::set<pg_shard_t>& acting);
+
+  // Replica constructor: no primary map
+  ScrubBackend(PgScrubber& scrubber,
+               PGBackend& backend,
+               PG& pg,
+               pg_shard_t i_am,
+               bool repair,
+               scrub_level_t shallow_or_deep);
+
+  friend class PgScrubber;
+
+  /**
+   * reset the per-chunk data structure (ScrubberBeChunk).
+   * Create an empty scrub-map for this shard, and place it
+   * in the appropriate entry in 'received_maps'.
+   *
+   * @returns a pointer to the newly created ScrubMap.
+   */
+  ScrubMap* new_chunk() final;
+
+  /**
+   * sets Backend's m_repair flag (setting m_mode_desc to a corresponding
+   * string)
+   */
+  void update_repair_status(bool should_repair) final;
```

Comment:
This only seems to get called during PGScrubber::scrub_finish.  I'm not sure I understand the intent.

Reply:
Basically - a copy of PgScrubber's mode text and 'repairing' status. Created so that the backend won't have to query 'backwards' for these details.